### PR TITLE
feat: ParseableType ModuleName supporting paths or module identifiers

### PR DESCRIPTION
**Pull request by semorrison**
We use this in Mathlib now, and might as well offer it upstream.

### CODE REVIEW

File: Cli/Basic.lean [hunk 1]
```diff
@@ -232,6 +232,17 @@ section Configuration
...
       | s  => s.toInt?
 
+  open Lean System in
+  /-- A custom command-line argument parser that allows either relative paths to Lean files,
+  (e.g. `Mathlib/Topology/Basic.lean`) or the module name (e.g. `Mathlib.Topology.Basic`). -/
+  instance : ParseableType Name where
+    name     := "Name"
+    parse? s :=
+      if s.endsWith ".lean" then
+        some <| (s : FilePath).withExtension "" |>.components.foldl Name.mkStr Name.anonymous
```
Comment (mhuisi) [inline Cli/Basic.lean:242]:
```suggestion
        let pathComponents := (s : FilePath).withExtension "" |>.components
        some <| pathComponents.foldl Name.mkStr Name.anonymous
```
?
Not a big fan of mixing `<|` and `|>` in the same line.

This should probably also account for `«...»`.

File: Cli/Basic.lean [hunk 2]
```diff
@@ -232,6 +232,17 @@ section Configuration
...
+      else
+        String.toName s
```
Comment (mhuisi) [inline Cli/Basic.lean:244]: Note that this yields `unreachable!` when given a digit and after that a literal that cannot be parsed to `Nat`, or `Name.num` when given a literal that parses to `Nat`. Both of these should likely yield `Option.none`.

Comment (semorrison) [reply Cli/Basic.lean:244]: Easy enough to handle the `Name.num` result.

It feels like the `unreachable!` in `String.toName` is an upstream bug, not sure what to do about that; I don't really want to re-implement the function here.

File: Cli/Basic.lean (hunk 1 above)
Comment (semorrison) [reply Cli/Basic.lean:242]: Sorry, I'm not certain what account for `«...»` you have in mind. This code trying to parse a path: no one puts `«...»` in their file names, do they?

File: Cli/Basic.lean (hunk 2 above)
Comment (mhuisi) [reply Cli/Basic.lean]: How about yielding `Option.none` if the first character is a digit? This would cover both cases.

It would also be nice to have some tests for this `ParseableType` instance in `Tests.lean`.

Comment (mhuisi) [reply Cli/Basic.lean]: Yeah, it looks like an upstream bug given that `String.toName` is also used in lots of user-facing code there. It's probably good enough to handle only the `Name.num` case then and perhaps fix it upstream some other time as well.

Comment (mhuisi) [reply Cli/Basic.lean]: Ah, it's not actually a bug. These kinds of names are mapped to `Name.anonymous` by `Lean.Syntax.splitNameLitAux`.

File: Cli/Basic.lean [hunk 3]
```diff
@@ -232,16 +232,26 @@ section Configuration
...
     parse? s :=
       if s.endsWith ".lean" then
-        some <| (s : FilePath).withExtension "" |>.components.foldl Name.mkStr Name.anonymous
+        let pathComponents := (s : FilePath).withExtension "" |>.components
+        some <| pathComponents.foldl Name.mkStr Name.anonymous
       else
-        String.toName s
+        match String.toName s with
+        | n@(.str _ _) => some n
+        | _ => none
```
Comment (mhuisi) [inline Cli/Basic.lean:254]: Note that this only considers the head of the name. How about the following:
```lean
  def ModuleName.isValid (m : ModuleName) : Bool :=
    match m with
    | .anonymous => false
    | _          => loop m
  where
    loop : ModuleName → Bool
      | .anonymous => true
      | .str pre _ => loop pre
      | .num ..    => false

  open Lean System in
  /-- A custom command-line argument parser that allows either relative paths to Lean files,
  (e.g. `Mathlib/Topology/Basic.lean`) or the module name (e.g. `Mathlib.Topology.Basic`). -/
  instance : ParseableType ModuleName where
    name     := "ModuleName"
    parse? s := do
      if s.endsWith ".lean" then
        let pathComponents := (s : FilePath).withExtension "" |>.components
        return pathComponents.foldl Name.mkStr Name.anonymous
      else
        let name := String.toName s
        guard <| ModuleName.isValid name
        return name
```

Comment (semorrison) [reply Cli/Basic.lean:254]: Looks good. I added a doc-string.

**Comment by mhuisi** [comment]
LGTM. I'll adjust some additional minor things after this is merged (handle some additional edge cases, add missing ModuleName instances, add tests, adjust README and `Example.lean`).
